skyportal/models.py

```diff
@@ -1645,39 +1652,170 @@ class Spectrum(Base):
     assignment_id = sa.Column(sa.ForeignKey('classicalassignments.id'), nullable=True)
     assignment = relationship('ClassicalAssignment', back_populates='spectra')
 
+    altdata = sa.Column(
+        psql.JSONB, doc="Miscellaneous alternative metadata.", nullable=True
+    )
+
+    original_file_string = sa.Column(
+        sa.String,
+        doc="Content of original file that was passed to upload the spectrum.",
+    )
+    original_file_filename = sa.Column(
+        sa.String, doc="Original file name that was passed to upload the spectrum."
+    )
+
     @classmethod
-    def from_ascii(cls, filename, obj_id, instrument_id, observed_at):
+    def from_ascii(
+        cls,
+        filename,
+        data=None,
+        obj_id=None,
+        instrument_id=None,
+        observed_at=None,
+        wave_colindex=0,
+        flux_colindex=1,
+        fluxerr_colindex=2,
+    ):
         """Generate a `Spectrum` from an ascii file.
 
         Parameters
         ----------
-
         filename : str
-           The name of the ASCII file containing the spectrum.
-        obj_id : str
-           The name of the Spectrum's Obj.
-        instrument_id : int
-           ID of the Instrument with which this Spectrum was acquired.
-        observed_at : string or datetime
-           Median UTC ISO time stamp of the exposure or exposures in which the Spectrum was acquired."
+           The original name of the ASCII file containing the spectrum,
+           or the name of the file on disk to read. If `data` is not provided,
+           `from_ascii` tries to read this file and parse its contents. If `data`
+           is provided, the filename is used only for bookkeeping purposes; no
+           attempt is made to actually read the file.
+        data: str, optional
```

Comment:
It would be simpler to avoid the joint handling of filename and data.  How about opting for `file` as a parameter, and allow that parameter to be a filename or file-like object.  If you want to read bytes, you put those into BytesIO and pass it in.

Reply:
Thanks, this is much nicer.